### ISSUE TITLE
CpuEngine.BatchNorm4D throws shape-vs-data mismatch on non-power-of-2 spatial dims (paper-faithful EfficientNet / MobileNet 224x224 chain blocked)

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#310]
## Summary

`CpuEngine.BatchNorm4D<T>` throws `ArgumentException: Data length (X) must match shape total (Y)` from inside `TensorAllocator.Rent[T](Int32[] shape, Vector<T> data)` whenever called with an input tensor whose spatial dimensions are not a power of two. The error fires at every batch size, not just batch=1.

This blocks paper-faithful inference on EfficientNet (Tan & Le 2019), MobileNetV2 (Sandler et al. 2018), and MobileNetV3 (Howard et al. 2019) — all three use the canonical 224 → 112 → 56 → 28 → 14 → 7 spatial chain that hits the bug at every stage where BN runs in training mode.

## Reproduction

```csharp
using AiDotNet.NeuralNetworks.Layers;
using AiDotNet.Tensors.LinearAlgebra;
using AiDotNet.ActivationFunctions;

var conv = new ConvolutionalLayer<float>(
    outputDepth: 32, kernelSize: 3, stride: 2, padding: 1,
    activationFunction: new IdentityActivation<float>());
var bn = new BatchNormalizationLayer<float>();
// LayerBase.IsTrainingMode defaults to true (matches PyTorch nn.Module's training=True).
// The eval-mode path through ApplyInferenceAnyRank works correctly — the bug is
// specifically in the training-mode BatchNorm4D dispatch.

var input = new Tensor<float>(new[] { 1, 3, 224, 224 });
for (int i = 0; i < input.Length; i++) input[i] = (float)((i % 7) * 0.1 - 0.3);

var convOut = conv.Forward(input);
// convOut.Shape = [1, 32, 112, 112], convOut.Length = 401408 — both correct
var bnOut = bn.Forward(convOut);
// THROWS: System.ArgumentException : Data length (524288) must match shape total (401408).
//   at AiDotNet.Tensors.Helpers.TensorAllocator.Rent[T](Int32[] shape, Vector<T> data)
//   at AiDotNet.Tensors.Engines.CpuEngine.BatchNorm4D[T](Tensor<T> input, Tensor<T> gamma, Tensor<T> beta, T eps, INumericOperations<T> numOps, Tensor<T>& mean, Tensor<T>& variance)
//   at AiDotNet.Tensors.Engines.CpuEngine.BatchNorm[T](Tensor<T> input, Tensor<T> gamma, Tensor<T> beta, Double epsilon, Tensor<T>& mean, Tensor<T>& variance)
```

## Failure pattern

The data length and shape total are consistently in a `next-power-of-2 spatial` ratio:

| Input shape           | shape total | data length | data factorization      |
|-----------------------|-------------|-------------|-------------------------|
| `[1, 32, 112, 112]`   | 401,408     | 524,288     | `32 * 128² = 524288`    |
| `[2, 32, 112, 112]`   | 802,816     | 1,048,576   | `2 * 32 * 128²`         |
| `[4, 32, 112, 112]`   | 1,605,632   | 2,097,152   | `4 * 32 * 128²`         |
| `[8, 32, 112, 112]`   | 3,211,264   | 4,194,304   | `8 * 32 * 128²`         |
| `[1, 96, 112, 112]`   | 1,204,224   | 2,097,152   | `128 * 128² = 2097152`  |

The data buffer is being sized to the next power of 2 in the spatial dim (112 → 128) and the channel dim (96 → 128), but the shape array stays at the logical extent. `TensorAllocator.Rent` enforces shape-total == data-length, so the call throws.

Tested at AiDotNet.Tensors version currently referenced by ooples/AiDotNet master (https://github.com/ooples/AiDotNet/blob/master/Directory.Packages.props).

## Working configurations

Power-of-2 spatial dims work fine:
- `[1, 16, 16, 16]` → OK
- `[1, 32, 64, 64]` → OK

The eval-mode BN path (`ApplyInferenceAnyRank` via `BatchNormalizationLayer.Forward` when `IsTrainingMode == false`) works correctly even on `[1, 32, 112, 112]`.

## Expected behavior

Per Ioffe & Szegedy 2015, BatchNorm in training mode should produce normalized output of the same shape as the input, regardless of spatial dim. The shape contract is logical-extent only.

## Likely fix

`BatchNorm4D` is calling `TensorAllocator.Rent(shape, data)` with the input's logical shape but the input's SIMD-padded data buffer. Either:
1. Compute mean/variance into a fresh tensor allocated with `new Tensor<T>(shape)` (no data sharing), or
2. Slice the input data down to `shape.Aggregate(*) == data.Length` before calling `Rent`.

## Why it matters for AiDotNet

EfficientNet B0 / B1 / ... / B7, MobileNetV2, MobileNetV3 cannot be trained or evaluated in their paper-default `training=true` state on the canonical input resolution. The downstream PR (ooples/AiDotNet#1224 cluster F) is currently blocked on:
- `EfficientNetTests.EfficientNetB0_Forward_ProducesCorrectOutputShape`
- `EfficientNetTests.EfficientNetB0_ForwardSmallInput_ProducesCorrectShape`
- `EfficientNetNetworkTests.Training_ShouldReduceLoss`
- `EfficientNetNetworkTests.LossStrictlyDecreasesOnMemorizationTask`
- `EfficientNetNetworkTests.MoreData_ShouldNotDegrade`
- `MobileNetV2NetworkTests.Training_ShouldReduceLoss`
- `MobileNetV2NetworkTests.LossStrictlyDecreasesOnMemorizationTask`
- `MobileNetV2NetworkTests.MoreData_ShouldNotDegrade`

We routed the unit tests through `Predict` (eval-mode) as a paper-faithful workaround for inference-only checks, but the training tests have no clean workaround.
